server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraJamesDefinition.java

```diff
@@ -0,0 +1,137 @@
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james;
+
+import static org.apache.james.CassandraJamesServerMain.ALL_BUT_JMX_CASSANDRA_MODULE;
+
+import java.util.List;
+
+import org.apache.james.mailbox.extractor.TextExtractor;
+import org.apache.james.mailbox.store.search.PDFTextExtractor;
+import org.apache.james.modules.TestESMetricReporterModule;
+import org.apache.james.modules.TestJMAPServerModule;
+import org.apache.james.server.core.configuration.Configuration;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Module;
+import com.google.inject.util.Modules;
+
+public class CassandraJamesDefinition implements JamesServerExtension.JamesDefinition {
+
+    public static class Builder {
+        private final ImmutableList.Builder<GuiceModuleTestExtension> extensions;
+        private final ImmutableList.Builder<Module> modules;
+
+        private Builder() {
+            extensions = ImmutableList.builder();
+            modules = ImmutableList.builder();
+        }
+
+        public Builder defaultExtensions() {
+            this.extensions.add(
+                new EmbeddedElasticSearchExtension(),
+                new CassandraRegistrableExtension());
+            return this;
+        }
+
+        public Builder addExtensions(GuiceModuleTestExtension... extensions) {
+            this.extensions.add(extensions);
+            return this;
+        }
+
+        public Builder addModules(Module... modules) {
+            this.modules.add(modules);
+            return this;
+        }
+
+        public CassandraJamesDefinition build() {
+            return new CassandraJamesDefinition(modules.build(), extensions.build());
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    static class CassandraRegistrableExtension implements GuiceModuleTestExtension {
+
+        private final DockerCassandraRule cassandra;
+
+        CassandraRegistrableExtension() {
+            this.cassandra = new DockerCassandraRule();
+        }
+
+        @Override
+        public void beforeAll(ExtensionContext extensionContext) {
+            cassandra.start();
+        }
+
+        @Override
+        public void afterAll(ExtensionContext extensionContext) {
+            cassandra.stop();
+        }
+
+        @Override
+        public Module getModule() {
+            return cassandra.getModule();
+        }
+    }
+
+    private static final int LIMIT_TO_10_MESSAGES = 10;
+
+    private final TemporaryFolderRegistrableExtension folderRegistrableExtension;
+    private final List<Module> additionalModules;
+    private final List<GuiceModuleTestExtension> extensions;
+
+    private CassandraJamesDefinition(List<Module> additionalModules, List<GuiceModuleTestExtension> extensions) {
```

Comment:
In my opinion, it doesn't make sense to introduce this CassandraJamesDefinition type.
It's a JamesServerExtension.JamesDefinition with some default values. These default values would be better located into the builder (it's what a builder is for, right ?).
Also, JamesDefinition looks useless too once you drop CassandraJamesDefinition.
The fact is : you want to build a JamesServerExtension with some values in it. Only JamesServerExtension has logic, everything else is configuration and doesn't deserve their own types. 
Having a builder building a type that is itself calling some builder should raise a warning.

Reply:
`Having a builder building a type that is itself calling some builder should raise a warning`

Not sure about this...

GuiceJamesServer builder is `generic` and `low level`. **CassandraServerDefinition** builder intends to be high level and test oriented.

I don't see how it hurts to provide the API needed for a given usage context...

Reply:
`Also, JamesDefinition looks useless too once you drop CassandraJamesDefinition`

Or it will be saving us as soon as we introduce another complex product with external dependencies that we dockerize...

Reply:
I already proposed an alternative to this PR here : https://github.com/linagora/james-project/pull/1766 but I will answer some questions here.

> GuiceJamesServer builder is generic and low level. CassandraServerDefinition builder intends to be high level and test oriented.
>
> I don't see how it hurts to provide the API needed for a given usage context...

GuiceJamesServer is not low level, it's the top-level object we use to run James in a few lines.

How it hurts to have another API ? If there are missing things in GuiceJamesServer, it's probably better to fix it right there instead of creating another abstraction. If GuiceJamesServer requires too many details, let's enhance it.

At this point, here is what's required to build a James server : 
```java
        GuiceJamesServer server = GuiceJamesServer.forConfiguration(configuration)
                    .combineWith(ALL_BUT_JMX_CASSANDRA_MODULE, new JMXServerModule());
```

I doubt it's too much.

Now you probably want to avoid repetition, for example to avoid copy/pasting this : 
```java
GuiceJamesServer.forConfiguration(configuration)
            .combineWith(ALL_BUT_JMX_CASSANDRA_MODULE)
            .overrideWith(Modules.combine(
                binder -> binder.bind(TextExtractor .class).to(PDFTextExtractor .class),
                new TestJMAPServerModule(LIMIT_TO_10_MESSAGES),
                new TestESMetricReporterModule()))
            .overrideWith(DOMAIN_LIST_CONFIGURATION_MODULE)
```
but nothing prevent you from putting that into a CONSTANT. you will even be able to add custom things on top of it.

On the other hand, if you hide the builder under a "better builder", you restrict the freedom of the caller for no benefit and you'll probably add methods to this builder to accommodate your test needs.